### ISSUE TITLE
Extendable API

### ISSUE THREAD

**Issue by 4shub** [glorious-codes/glorious-demo#58]
I was wondering if there were plans to create an extensible API so users can create different types of windows. I would be very interested in contributing, as I am working on a fork of this project to allow users to demo a web browser experience as well.

I wouldn't want that browser window to be added to this application as it would bloat the app unnecessarily for most people, but I think lots would value if they could add it as an option if we added some sort of API to add extra types of applicatons.

**Comment by rafaelcamargo** [comment]
This is a nice suggestion. I myself already have been thinking about this exactly by the need to demonstrate a couple of action on a browser.

However, I confess that I have not thought any way of implementing this. The plugin strategy seems very interesting to me. If you like to propose some implementation strategy, please go ahead. It will be a pleasure to me to discuss that 👍

**Comment by 4shub** [comment]
Sure, I was thinking of making a very simple extension API that acts unobtrusively with the existing API. 

Something like the following:
```
import GDemo, { extender } from '@glorious/demo';

extender({ name: 'browser', content: BrowserApplication });
```

This extender would ultimately just add an application type to some list we create in `desktop.js`

```
const ApplicationTypes = {
  editor: EditorApplication,
  terminal: TerminalApplication,
};

export const extender = (newApplicationType) => {
  ApplicationTypes[newApplicationType.name] = newApplicationType.content;
};
``` 

This way I feel is a little better than just invoking in let's say `new GDemo('selector', options`) because we would assume if a user wanted to extend their gDemo with new application types, they would probably want to just set it once and be done with it.

**Comment by rafaelcamargo** [comment]
Interesting approach. Two things that I liked:

1. Clear and simple to implement.
2. In fact, seems to solve peculiarities that are beyond the scope of the lib.

Some concerns:

1. `openApp` starts to be used differently depending on the app to be open.
2. The specific app just created does not inherit features that are related to an [Application](https://github.com/glorious-codes/glorious-demo/blob/master/src/scripts/components/application/application.js). e.g.: windowTitle, minHeight, etc. This may break expectations related to the acceptable options.

What about the following?

``` javascript
import GDemo from '@glorious/demo';
import browserApp from 'gdemo-browser-app'; // client implementation or a package published on NPM

GDemo.addApp(browserApp);

const gDemo = new GDemo('#selector');

gdemo
  .openApp('browser', { minHeight: '400px', windowTitle: 'Github' })
  .loadPage(...) // example of app specific method
  .end();
```

That way, the implementation of browserApp must extend the base Application:

``` javascript
// gdemo-browser-app.js
import { Application } from '@glorious/demo';

export class BrowserApplication extends Application {
  constructor(options) {
    super('browser', options);
  }
  loadPage() {
    // app specific method 
  }
}
```

The changes on `Desktop.js` seems to be very close to the changes you have suggested. Perhaps, we need to create a service to retrieve applications by application name from an application enum.

What do you think?

**Comment by 4shub** [comment]
That looks good to me, and I agree that we should have some sort of service to retrieve these applications because I realized that we would also have to extend `index.js` to allow for users to add new custom-app-specific methods. 

We would have to do a small rewrite where we have to force the user to invoke `openApp` before running any method that isn't `openApp`, and make the `app` attribute in the steps just rely on the last called 'openApp' to determine what app we need to run the application specific method from/

**Comment by rafaelcamargo** [comment]
Right, I see your point. Well, below is how exactly works the flow of opening/closing applications:

1. First of all, every animation step is registered in an Array containing the details of each step. At this phase, no animation is still playing.
2. When user invokes `.end()`, the array of steps is passed to the [Player](https://github.com/glorious-codes/glorious-demo/blob/master/src/scripts/components/player/player.js) which is responsible to [play](https://github.com/glorious-codes/glorious-demo/blob/master/src/scripts/components/player/player.js#L10) every step following its details.
3. The Player instantiate a Desktop. The Desktop is the only component that manages applications. It builds, maximizes (opens) or minimizes applications when Player [asks it](https://github.com/glorious-codes/glorious-demo/blob/master/src/scripts/components/player/player.js#L39).
4. When Desktop brings the application to the Player, Player [executes](https://github.com/glorious-codes/glorious-demo/blob/master/src/scripts/components/player/player.js#L41) the application method informed in the current step.

The problem is that every application method is interfaced through GDemo class methods. If GDemo class has no custom application methods, the steps for this custom application couldn't be registered.

That said, I have some concerns on your approach:
> I realized that we would also have to extend index.js

1. If user extends GDemo class to add custom application methods to it, we take the risk of having method name collision - one plugin could potentially break another 😢 

So, what if the custom application (plugin) returned its own interface to register steps and added them to the steps on the instance of [GDemo](https://github.com/glorious-codes/glorious-demo/blob/master/src/scripts/index.js#L8)?

I am figuring out here how to make this in the Custom Application (plugin). As soon as I have one first solution, I put it here 👍

**Comment by rafaelcamargo** [comment]
Take this as a simple brainstorm. I confess that I still don't like the way some Custom App is being implemented (seems too complex).

``` javascript
import GDemo from '@glorious/demo';
import BrowserApp from 'gdemo-browser-app'; // client implementation or a package published on NPM

GDemo.addCustomApp(BrowserApp);

const gDemo = new GDemo('#selector');

gdemo
  .openApp('browser', { minHeight: '400px', windowTitle: 'Github' })
  .loadPage(...) // example of app specific method
  .end();
```

``` javascript
// index.js
import customApplicationsService from './services/custom-applications/custom-applications';

export default class {
  static addCustomApp = function(App) {
    customApplicationsService.add(App);
  }
  constructor(selector){
    this.container = document.querySelector(selector);
    this.customApps = [];
    this.steps = [];
  }
  openApp(app, options = {}){
    if(isBuiltInApp(app)) {
      this.steps.push({
        app,
        options,
        onCompleteDelay: options.onCompleteDelay
      });
      return this;
    } else {
      return getCustomApp(app, options, this);
    }
  }
  ...
}

function isBuiltInApp(app) {
  return ['editor', 'terminal'].includes(app);
}

function getCustomApp(app, options, gDemo) {
  // Look up on this.customApps. If found it, return it. Otherwise:
  const Application = customApplicationsService.find(app);
  const customApp = new Application(gDemo.steps, options);
  buildProxyForGDemoOpenAppInCustomApp(customApp, gDemo)
  return customApp;
}

function buildProxyForGDemoOpenAppInCustomApp(customApp, gDemo) {
  customApp.openApp = function(app, options) {
     return gDemo.openApp(app, options);
  };
}
```

``` javascript
// gdemo-browser-app.js
import { Application } from '@glorious/demo';

const app = 'browser';

export class BrowserApplication extends Application {
  static app= app;
  constructor(steps, options) {
    super(app, options);
    this.steps = steps;
  }
  loadPage(mainParams, options) {
    this.steps.push({
      app,
      action: '_loadPage',
      params: mainParams,
      onCompleteDelay: options.onCompleteDelay
    });
    return this;
  }
  _loadPage(params) {
    // some implementation
  }
}
```
